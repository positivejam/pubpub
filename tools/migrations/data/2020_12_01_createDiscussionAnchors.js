/* eslint-disable no-console */
import { Anchor, Branch, Pub, Release, Discussion } from 'server/models';
import { getBranchRef } from 'server/utils/firebaseAdmin';
import { uncompressSelectionJSON } from 'prosemirror-compress-pubpub';

import { forEach, forEachInstance } from '../util';

const getAnchorDescsFromSingleFirebaseDiscussion = (firebaseDiscussion, isPublicBranch) => {
	if (firebaseDiscussion) {
		const {
			initKey,
			initHead,
			initAnchor,
			currentKey,
			selection: compressedSelection,
		} = firebaseDiscussion;
		const selection = uncompressSelectionJSON(compressedSelection);
		return [
			{
				historyKey: initKey,
				head: initHead,
				anchor: initAnchor,
				isPublicBranch: isPublicBranch,
			},
			{
				historyKey: currentKey,
				head: selection.head,
				anchor: selection.anchor,
				isPublicBranch: isPublicBranch,
			},
		];
	}
	return [];
};
const getAnchorDescsFromFirebaseDiscussions = ({
	discussions,
	firebaseDiscussions,
	isPublicBranch,
}) => {
	return discussions
		.map((discussion) =>
			getAnchorDescsFromSingleFirebaseDiscussion(
				firebaseDiscussions[discussion.id],
				isPublicBranch,
			),
		)
		.reduce((a, b) => [...a, ...b], []);
};

const mapAnchorDescsToDraft = (anchorDescs, releases) => {
	return anchorDescs.map((anchor) => {
		const { isPublicBranch, historyKey } = anchor;
		if (isPublicBranch) {
			const correspondingRelease = releases[Math.min(historyKey, releases.length - 1)];
			const draftKey = correspondingRelease.historyKey;
			return { ...anchor, historyKey: draftKey };
		}
		return anchor;
	});
};

const handlePub = async (pub) => {
	const [branches, releases, discussions] = await Promise.all([
		Branch.findAll({ where: { pubId: pub.id } }),
		Release.findAll({ where: { pubId: pub.id } }),
		Discussion.findAll({
			where: { pubId: pub.id },
			include: [{ model: Anchor, as: 'anchor' }],
		}),
	]);
	const anchorDescs = [];
	const publicBranch = branches.find((br) => br.title === 'public');
	// Retrieve all anchors from Firebase
	await forEach(branches, async (branch) => {
		const branchRef = getBranchRef(pub.id, branch.id);
		const discussionsSnapshot = await branchRef.child('discussions').once('value');
		const firebaseDiscussions = discussionsSnapshot.val();
		const isPublicBranch = branch === publicBranch;
		const branchAnchorDescs = getAnchorDescsFromFirebaseDiscussions({
			discussions: discussions,
			firebaseDiscussions: firebaseDiscussions,
			isPublicBranch: isPublicBranch,
		});
		anchorDescs.push(...branchAnchorDescs);
	});
	// Get all anchors from associated discussion.anchor models
	const anchorModelDescs = discussions
		.map((discussion) => {
			const { anchor } = discussion;
			if (anchor) {
				const { from, to, prefix, suffix, exact, branchId, branchKey } = anchor;
				return {
					anchor: from,
					head: to,
					prefix: prefix,
					suffix: suffix,
					exact: exact,
					isPublicBranch: branchId === publicBranch.id,
					historyKey: branchKey,
				};
			}
			return null;
		})
		.filter((x) => x);
	anchorDescs.push(...anchorModelDescs);
	const draftMappedAnchorDescs = mapAnchorDescsToDraft(anchorDescs, releases);
	console.log(pub.title, draftMappedAnchorDescs);
};

module.exports = async () => {
	await forEachInstance(Pub, handlePub);
};
