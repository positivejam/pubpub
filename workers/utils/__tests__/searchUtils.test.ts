/* global describe, it, expect, beforeAll, afterAll */
import { stub, modelize, setup, teardown, determinize } from 'stubstub';
import { plainDoc, fullDoc } from 'utils/storybook/data';

import { getPageSearchData, getPubSearchData } from '../searchUtils';

const models = modelize`
    Community {
        title: "Test Community"
        subdomain: "tc1"
        Pub shortPub {
            title: "A Short Pub"
            slug: "a-short-pub"
            description: "I am very brief"
            avatar: "short.png"
            PubAttribution {
                User {
                    slug: "test1"
                }
            }
        }
        Pub longPub {
            title: "A Long Pub"
            slug: "a-long-pub"
            description: "I am very lengthy"
            avatar: "long.png"
            PubAttribution {
                isAuthor: false
                order: 0.3
                User {
                    slug: "test3"
                }
            }
            PubAttribution {
                isAuthor: true
                order: 0.2
                User {
                    firstName: "First"
                    lastName: "Author"
                    fullName: "Second Author"
                    slug: "test4"
                }
            }
            PubAttribution {
                isAuthor: true
                order: 0.1
                User {
                    firstName: "First"
                    lastName: "Author"
                    fullName: "First Author"
                    slug: "test2"
                }
            }
        }
        Page testPage {
            title: "I am a test page"
            slug: "test-page"
            description: "You can put text or Pubs on me"
            avatar: "page.png"
        }
    }
`;

let firebaseStub;

setup(beforeAll, async () => {
	await models.resolve();
	firebaseStub = stub('server/utils/firebaseAdmin', {
		getBranchDoc: (pubId) => {
			return Promise.resolve({ doc: pubId === models.longPub.id ? fullDoc : plainDoc });
		},
	});
});

teardown(afterAll, () => {
	firebaseStub.restore();
});

const determinizePubData = determinize([
	'avatar',
	'branchIsPublic',
	'branchShortId',
	'byline',
	'communityDomain',
	'communityTitle',
	'description',
	'slug',
	'title',
]);

const determinizePageData = determinize([
	'avatar',
	'communityDomain',
	'communityTitle',
	'description',
	'slug',
	'title',
]);

describe('getPubSearchData', () => {
	it('produces the expected data for a long Pub', async () => {
		const data = await getPubSearchData(models.longPub.id);
		expect(data.map(determinizePubData)).toMatchSnapshot();
	});

	it('produces the expected data for a short Pub', async () => {
		const data = await getPubSearchData(models.shortPub.id);
		expect(data.map(determinizePubData)).toMatchSnapshot();
	});
});

describe('getPageSearchData', () => {
	it('produces the expected data for a Page', async () => {
		const data = await getPageSearchData(models.testPage.id);
		expect(data.map(determinizePageData)).toMatchSnapshot();
	});
});
