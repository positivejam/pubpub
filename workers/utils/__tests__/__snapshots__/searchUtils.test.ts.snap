// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getPageSearchData produces the expected data for a Page 1`] = `
Array [
  Object {
    "avatar": "page.png",
    "communityDomain": "tc1.pubpub.org",
    "communityTitle": "Test Community",
    "description": "You can put text or Pubs on me",
    "slug": "test-page",
    "title": "I am a test page",
  },
]
`;

exports[`getPubSearchData produces the expected data for a long Pub 1`] = `
Array [
  Object {
    "avatar": "long.png",
    "branchIsPublic": true,
    "branchShortId": 1,
    "byline": "by First Author, Second Author, and Test Testington",
    "communityDomain": "tc1.pubpub.org",
    "communityTitle": "Test Community",
    "description": "I am very lengthy",
    "slug": "a-long-pub",
    "title": "A Long Pub",
  },
  Object {
    "avatar": "long.png",
    "branchIsPublic": false,
    "branchShortId": 2,
    "byline": "by First Author, Second Author, and Test Testington",
    "communityDomain": "tc1.pubpub.org",
    "communityTitle": "Test Community",
    "description": "I am very lengthy",
    "slug": "a-long-pub",
    "title": "A Long Pub",
  },
]
`;

exports[`getPubSearchData produces the expected data for a short Pub 1`] = `
Array [
  Object {
    "avatar": "short.png",
    "branchIsPublic": true,
    "branchShortId": 1,
    "byline": "by Test Testington",
    "communityDomain": "tc1.pubpub.org",
    "communityTitle": "Test Community",
    "description": "I am very brief",
    "slug": "a-short-pub",
    "title": "A Short Pub",
  },
  Object {
    "avatar": "short.png",
    "branchIsPublic": false,
    "branchShortId": 2,
    "byline": "by Test Testington",
    "communityDomain": "tc1.pubpub.org",
    "communityTitle": "Test Community",
    "description": "I am very brief",
    "slug": "a-short-pub",
    "title": "A Short Pub",
  },
]
`;
